### PR TITLE
GCG-44 add support for q18 and partially q4

### PR DESCRIPTION
**Pull request by wqc0712**

### CODE REVIEW

File: codegen/src/test/resources/q4/Job.scala [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+      val ordersS = orders.keyBy(i => i._3)
+      .process(new Q4OrdersProcessFunction())
+      val lineitemS = ordersS.connect(lineitem)
+      .keyBy(i => i._3, i => i._3)
+      .process(new Q4LineitemProcessFunction())
+      val result = lineitemS.keyBy(i => i._3)
+      .map(x => Payload("Aggregate", "Addition", x._3, x._4, x._5, x._6))
+      .keyBy(i => i._3)
+      .process(new Q4Aggregate2ProcessFunction())
+      .writeAsText(outputpath,FileSystem.WriteMode.OVERWRITE)
```
Comment (wqc0712) [inline codegen/src/test/resources/q4/Job.scala:31]: There should be a .map(x => (x._4.mkString(", "), x._5.mkString(", "), x._6)) after line 30.

File: codegen/src/test/resources/q4/Job.scala [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+      val inputStream : DataStream[Payload] = getStream(env,inputpath)
+      val orders : DataStream[Payload] = inputStream.getSideOutput(ordersTag)
+      val lineitem : DataStream[Payload] = inputStream.getSideOutput(lineitemTag)
+      val ordersS = orders.keyBy(i => i._3)
+      .process(new Q4OrdersProcessFunction())
+      val lineitemS = ordersS.connect(lineitem)
+      .keyBy(i => i._3, i => i._3)
+      .process(new Q4LineitemProcessFunction())
+      val result = lineitemS.keyBy(i => i._3)
+      .map(x => Payload("Aggregate", "Addition", x._3, x._4, x._5, x._6))
```
Comment (wqc0712) [inline codegen/src/test/resources/q4/Job.scala:28]: The aggregate process function 1 is not used in the main class.  There should be a .process(new Q4Aggregate1ProcessFunction) after line 28.

File: codegen/src/test/resources/q4/Q4Aggregate2ProcessFunction.scala [hunk 3]
```diff
@@ -0,0 +1,19 @@
+import org.hkust.RelationType.Payload
+import org.apache.flink.api.common.state.ValueStateDescriptor
+import org.apache.flink.api.common.typeinfo.{TypeHint, TypeInformation}
+import org.hkust.BasedProcessFunctions.AggregateProcessFunction
+class Q4Aggregate2ProcessFunction extends AggregateProcessFunction[Any, Integer]("Q4Aggregate2ProcessFunction", Array("O_ORDERPRIORITY"), Array("O_ORDERPRIORITY"), aggregateName = "order_count", deltaOutput = true) {
+   override def aggregate(value: Payload): Integer = {
+      if(value("DIST_COUNT").asInstanceOf[String].toInt==1) 1
+if(value("DIST_COUNT").asInstanceOf[String].toInt==0) -1
```
Comment (wqc0712) [inline codegen/src/test/resources/q4/Q4Aggregate2ProcessFunction.scala:8]: Line 8 should have "else" at the beginning, makes the query an if-else { if-else } query.

File: codegen/src/test/resources/q4/Q4Aggregate2ProcessFunction.scala [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+else 0
+   }
+   override def addition(value1: Integer, value2: Integer): Integer = value1 + value2
+   override def subtraction(value1: Integer, value2: Integer): Integer = value1 - value2
+   override def initstate(): Unit = {
+      val valueDescriptor = TypeInformation.of(new TypeHint[Integer](){})
+      val aliveDescriptor : ValueStateDescriptor[Integer] = new ValueStateDescriptor[Integer]("Q4Aggregate2ProcessFunction"+"Alive", valueDescriptor)
+      alive = getRuntimeContext.getState(aliveDescriptor)
+      }
+         override val init_value: Integer = 0.0
```
Comment (wqc0712) [inline codegen/src/test/resources/q4/Q4Aggregate2ProcessFunction.scala:18]: The initial value for Integer should be 0.

File: codegen/src/test/resources/q4/Q4Aggregate1ProcessFunction.scala [hunk 5]
```diff
@@ -0,0 +1,17 @@
+import org.hkust.RelationType.Payload
+import org.apache.flink.api.common.state.ValueStateDescriptor
+import org.apache.flink.api.common.typeinfo.{TypeHint, TypeInformation}
+import org.hkust.BasedProcessFunctions.AggregateProcessFunction
+class Q4Aggregate1ProcessFunction extends AggregateProcessFunction[Any, Integer]("Q4Aggregate1ProcessFunction", Array("ORDERKEY"), Array("O_ORDERPRIORITY"), aggregateName = "DIST_COUNT", deltaOutput = true) {
```
Comment (wqc0712) [inline codegen/src/test/resources/q4/Q4Aggregate1ProcessFunction.scala:5]: Add nextKey : {O_ORDERPRIORITY} into the json file in AggregateProcessFunction 1, 

which will change this line to 

class Q4Aggregate1ProcessFunction extends AggregateProcessFunction[Any, Integer]("Q4Aggregate1ProcessFunction", Array("ORDERKEY"), Array("O_ORDERPRIORITY"), nextKey=Array("O_ORDERPRIORITY"), aggregateName = "DIST_COUNT", deltaOutput = true) {

Comment (DanishAlsayed) [reply codegen/src/test/resources/q4/Q4Aggregate1ProcessFunction.scala:5]: I thought there is no next_key anymore in aggregate process functions?

**Pull request merged**